### PR TITLE
Performance monitoring plugin intel_pmu

### PR DESCRIPTION
**Pull request by serhiyx**
The intel_pmu plugin reads performance counters provided by the Linux kernel perf interface. The plugin uses jevents library to resolve named events to perf events and access perf interface.

### CODE REVIEW

File: src/collectd.conf.pod [hunk 1]
```diff
@@ -3082,6 +3082,86 @@ Defaults to B<false>.
...
+    KernelPMUEvents true
+    SWEvents true
+    HWSpecificEvents "L2_RQSTS.CODE_RD_HIT,L2_RQSTS.CODE_RD_MISS"
+  </Plugin>
+
+B<Options:>
+
+=over 4
+
+=item B<HWCacheEvents> B<false>|B<true>
```
Comment (octo) [inline src/collectd.conf.pod]: Configuration options with this functionality are usually called `ReportFoo`, for example `ReportNumCpu` of the *CPU plugin*.

File: src/collectd.conf.pod [hunk 2]
```diff
@@ -3082,6 +3082,86 @@ Defaults to B<false>.
...
+  - dTLB-stores
+  - dTLB-store-misses
+  - dTLB-prefetches
+  - dTLB-prefetch-misses
+  - iTLB-loads
+  - iTLB-load-misses
+  - branch-loads
+  - branch-load-misses
+
+=item B<KernelPMUEvents> B<false>|B<true>
```
Comment (octo) [inline src/collectd.conf.pod]: Dito: `ReportKernelPMUEvents`

File: src/collectd.conf.pod [hunk 3]
```diff
@@ -3082,6 +3082,86 @@ Defaults to B<false>.
...
+Enable or disable measuring of the following events:
+  - cpu-cycles
+  - instructions
+  - cache-references
+  - cache-misses
+  - branches
+  - branch-misses
+  - bus-cycles
+
+=item B<SWEvents> B<false>|B<true>
```
Comment (octo) [inline src/collectd.conf.pod]: Dito: `ReportSWEvents`; please consider expanding "SW" to "Software" to make this option easier to read.

File: src/collectd.conf.pod [hunk 4]
```diff
@@ -3082,6 +3082,86 @@ Defaults to B<false>.
...
+  - cpu-migrations
+  - page-faults
+  - minor-faults
+  - major-faults
+  - alignment-faults
+  - emulation-faults
+
+=item B<HWSpecificEvents> I<events>
+
+This field is a list of comma separated event names. To be able to monitor all
```
Comment (octo) [inline src/collectd.conf.pod]: > comma separated

Config options can take multiple arguments. Please take a list of strings instead of a single string that is then parsed inside the plugin.

File: src/collectd.conf.pod [hunk 5]
```diff
@@ -3082,6 +3082,86 @@ Defaults to B<false>.
...
+  - task-clock
+  - context-switches
+  - cpu-migrations
+  - page-faults
+  - minor-faults
+  - major-faults
+  - alignment-faults
+  - emulation-faults
+
+=item B<HWSpecificEvents> I<events>
```
Comment (octo) [inline src/collectd.conf.pod]: With the boolean options all having the form "Report… true|false", I'd suggest to rename this option to **HardwareEvents**.

File: src/intel_pmu.c [hunk 6]
```diff
@@ -0,0 +1,512 @@
...
+ * SOFTWARE.
+ *
+ * Authors:
+ *   Serhiy Pshyk <serhiyx.pshyk@intel.com>
+ **/
+
+#include "collectd.h"
+#include "common.h"
+
+#include "jevents.h"
```
Comment (octo) [inline src/intel_pmu.c]: Please use angle brackets for files not included with collectd: `#include <jevents.h>`. Dito for `<jsession.h>`.

File: src/intel_pmu.c [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+
+  DEBUG(PMU_PLUGIN ": Events:");
+
+  struct event *e;
+
+  for (e = g_ctx.event_list->eventlist; e; e = e->next) {
+    DEBUG(PMU_PLUGIN ":   event       : %s", e->event);
+    DEBUG(PMU_PLUGIN ":     group_lead: %d", e->group_leader);
+    DEBUG(PMU_PLUGIN ":     end_group : %d", e->end_group);
+    DEBUG(PMU_PLUGIN ":     type      : 0x%X", e->attr.type);
```
Comment (octo) [inline src/intel_pmu.c]: Please use the `%#X` format instead of writing out the `0x` manually. Here and below.

File: src/intel_pmu.c [hunk 8]
```diff
@@ -0,0 +1,512 @@
...
+    DEBUG(PMU_PLUGIN ":     config    : 0x%X", (int)e->attr.config);
```
Comment (octo) [inline src/intel_pmu.c]: Please cast to `(unsigned)` when using the `%X` format.

File: src/intel_pmu.c [hunk 9]
```diff
@@ -0,0 +1,512 @@
...
+
+static void pmu_submit_counter(int cpu, char *event, counter_t value) {
+  value_list_t vl = VALUE_LIST_INIT;
+
+  vl.values = &(value_t){.counter = value};
+  vl.values_len = 1;
+
+  sstrncpy(vl.plugin, PMU_PLUGIN, sizeof(vl.plugin));
+  if (cpu == -1) {
+    snprintf(vl.plugin_instance, sizeof(vl.plugin_instance), "all");
```
Comment (octo) [inline src/intel_pmu.c]: Please use collectds wrapper, `ssnprintf()`. Here and below.

File: src/intel_pmu.c [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+
+  return (0);
+}
+
+static int pmu_add_events(struct eventlist *el, uint32_t type,
+                          event_info_t *events, int count) {
+
+  for (int i = 0; i < count; i++) {
+    struct event *e =
+        calloc(sizeof(struct event) + sizeof(struct efd) * el->num_cpus, 1);
```
Comment (octo) [inline src/intel_pmu.c]: This way of calling `calloc` defeats its purpose. Please use this instead:

```c
struct event *e = calloc(el->num_cpus, sizeof(*e) + sizeof(struct efd);
```

Please add a comment explaining why the `sizeof(struct efd)` is needed.

File: src/intel_pmu.c [hunk 11]
```diff
@@ -0,0 +1,512 @@
...
+    e->event = strdup(events[i].name);
+  }
+
+  return (0);
+}
+
+static int pmu_parse_events(struct eventlist *el, char *events) {
+  char *s, *tmp;
+
+  events = strdup(events);
```
Comment (octo) [inline src/intel_pmu.c]: Please assign this value to a new variable and change the argument to `char const *`.

File: src/intel_pmu.c [hunk 12]
```diff
@@ -0,0 +1,512 @@
...
+  if (!events)
+    return -1;
+
+  for (s = strtok_r(events, ",", &tmp); s; s = strtok_r(NULL, ",", &tmp)) {
+    bool group_leader = false, end_group = false;
```
Comment (octo) [inline src/intel_pmu.c]: Please use `_Bool` and `0` / `1` instead of the `<stdbool.h>` aliases.

File: src/intel_pmu.c [hunk 13]
```diff
@@ -0,0 +1,512 @@
...
+    if (s[0] == '{') {
+      s++;
+      group_leader = true;
+    } else if (len = strlen(s), len > 0 && s[len - 1] == '}') {
+      s[len - 1] = 0;
+      end_group = true;
+    }
+
+    struct event *e =
+        calloc(sizeof(struct event) + sizeof(struct efd) * el->num_cpus, 1);
```
Comment (octo) [inline src/intel_pmu.c]: As above: `calloc(el->num_cpus, sizeof(*e) + sizeof(struct efd))`

File: src/intel_pmu.c [hunk 14]
```diff
@@ -0,0 +1,512 @@
...
+    if (e == NULL) {
+      ERROR(PMU_PLUGIN ": Failed to allocate event structure");
+      return (-ENOMEM);
+    }
+
+    e->attr.type = type;
+    e->attr.config = events[i].config;
+    e->attr.size = PERF_ATTR_SIZE_VER0;
+    e->group_leader = false;
```
Comment (octo) [inline src/intel_pmu.c]: Please don't use the `false` define, instead use `0`. Since `calloc` zeroes the memory, you can just remote this and the following line.

File: src/intel_pmu.c [hunk 15]
```diff
@@ -0,0 +1,512 @@
...
+    if (el->eventlist_last)
+      el->eventlist_last->next = e;
+    el->eventlist_last = e;
+    e->event = strdup(events[i].name);
+  }
+
+  return (0);
+}
+
+static int pmu_parse_events(struct eventlist *el, char *events) {
```
Comment (octo) [inline src/intel_pmu.c]: As mentioned above: the config parser can give you a list of strings – please use that instead of implementing your own (sub-)parsing.

File: src/intel_pmu.c [hunk 16]
```diff
@@ -0,0 +1,512 @@
...
+    if (ret != 0) {
+      ERROR(PMU_PLUGIN ": Failed to setup perf events for the event list.");
+      goto init_error;
+    }
+  } else {
+    WARNING(PMU_PLUGIN
+            ": Events list is empty. No events were setup for monitoring.");
+  }
+
+  return (0);
```
Comment (octo) [inline src/intel_pmu.c]: Please unify this style. Both, `return foo;` and `return (foo);` are okay, just use the same everywhere in the file.

File: src/intel_pmu.c [hunk 17]
```diff
@@ -0,0 +1,512 @@
...
+static int pmu_parse_events(struct eventlist *el, char *events) {
+  char *s, *tmp;
+
+  events = strdup(events);
+  if (!events)
+    return -1;
+
+  for (s = strtok_r(events, ",", &tmp); s; s = strtok_r(NULL, ",", &tmp)) {
+    bool group_leader = false, end_group = false;
+    int len;
```
Comment (octo) [inline src/intel_pmu.c]: This should be a `size_t`.

File: src/intel_pmu.c [hunk 18]
```diff
@@ -0,0 +1,512 @@
...
+}
+
+static int pmu_read(__attribute__((unused)) user_data_t *ud) {
+  int ret;
+
+  DEBUG(PMU_PLUGIN ": %s:%d", __FUNCTION__, __LINE__);
+
+  ret = read_all_events(g_ctx.event_list);
+  if (ret != 0) {
+    DEBUG(PMU_PLUGIN ": Failed to read values of all events.");
```
Comment (octo) [inline src/intel_pmu.c]: This should be an `ERROR()` since you abort and return prematurely.

File: src/intel_pmu.c [hunk 19]
```diff
@@ -0,0 +1,512 @@
...
+    return (0);
+  }
+
+  ret = pmu_dispatch_data();
+  if (ret != 0) {
+    DEBUG(PMU_PLUGIN ": Failed to dispatch event values.");
```
Comment (octo) [inline src/intel_pmu.c]: Dito: Use `ERROR()` here.

File: src/intel_pmu.c (hunk 16 above)
Comment (rpv-tomsk) [inline src/intel_pmu.c]: @octo 
There is a WIP about parentheses removing (#2277). 
Maybe syntax with parentheses should be noted as forbidden / deprecated?

Comment (octo) [inline src/intel_pmu.c]: Agreed, with #2277, please remove all parenthesis from this PR.

File: src/intel_pmu.c (hunk 10 above)
Comment (serhiyx) [inline src/intel_pmu.c]: Comment added, but size did not change because I need  **sizeof(event) + sizeof(efd)*num_cpus** but not **(sizeof(event) + sizeof(efd)) * num_cpus**

File: src/intel_pmu.c (hunk 15 above)
Comment (serhiyx) [inline src/intel_pmu.c]: Refactored this function to receive list of strings provided by config parser. But some subparsing routine is still needed as each entry can contain multiple event names to create group of events.

File: src/collectd.conf.pod (hunk 4 above)
Comment (serhiyx) [inline src/collectd.conf.pod]: Refactored this functionality to use list of strings provided by config parser. But some subparsing routine is still needed as each entry can contain multiple event names to create group of events.

File: contrib/systemd.collectd.service [hunk 20]
```diff
@@ -11,6 +11,20 @@ EnvironmentFile=-/etc/default/collectd
...
+#
+# Here's a (incomplete) list of the plugins known environment requirements:
+#   intel_pmu       XDG_CACHE_HOME
+#
+# Example, if you use the intel_pmu plugin and cache is located in the /opt
+# directory:
+# Environment=XDG_CACHE_HOME=/opt
+#
+# By default, the HOME directory is chosen:
+Environment=
```
Comment (rubenk) [inline contrib/systemd.collectd.service]: Instead of getting this from the environment, would it be possible to turn this into a config option settable in collectd.conf?

File: configure.ac [hunk 21]
```diff
@@ -4364,6 +4364,77 @@ then
 fi
 # }}}
 
+# --with-libjevents {{{
+with_libjevents_cppflags=""
+with_libjevents_ldflags=""
+AC_ARG_WITH(libjevents, [AS_HELP_STRING([--with-libjevents@<:@=PREFIX@:>@], [Path to libjevents.])],
```
Comment (rubenk) [inline configure.ac]: Please keep the same code style as the rest of configure.ac

File: configure.ac [hunk 22]
```diff
@@ -4364,6 +4364,77 @@ then
...
+  SAVE_CPPFLAGS="$CPPFLAGS"
+  CPPFLAGS="$CPPFLAGS $with_libjevents_cppflags"
+
+  AC_CHECK_HEADERS(jevents.h, [with_libjevents="yes"], [with_libjevents="no (jevents.h not found)"])
+
+  CPPFLAGS="$SAVE_CPPFLAGS"
+fi
+if test "x$with_libjevents" = "xyes"
+then
+  SAVE_CPPFLAGS="$CPPFLAGS"
```
Comment (rubenk) [inline configure.ac]: AC_CHECK_LIB doesn't use CPPFLAGS, so no need to save them.

File: configure.ac [hunk 23]
```diff
@@ -4364,6 +4364,77 @@ then
...
+  CPPFLAGS="$SAVE_CPPFLAGS"
+  LDFLAGS="$SAVE_LDFLAGS"
+fi
+if test "x$with_libjevents" = "xyes"
+then
+  SAVE_CPPFLAGS="$CPPFLAGS"
+  SAVE_LDFLAGS="$LDFLAGS"
+  SAVE_LIBS="$LIBS"
+  CPPFLAGS="$CPPFLAGS -fPIC"
+  LDFLAGS="$LDFLAGS -shared"
```
Comment (rubenk) [inline configure.ac]: Setting -fPIC and -shared is not needed, you're creating an executable, not a library.

File: configure.ac [hunk 24]
```diff
@@ -4364,6 +4364,77 @@ then
...
+  AC_LINK_IFELSE([AC_LANG_SOURCE(
+    [[
+      #include <stdio.h>
+      #include "jevents.h"
+      void print_cpu(void){
+        printf("%s", get_cpu_str());
+      }
+    ]]
+  )],
+  [with_libjevents="yes"], [with_libjevents="no (could not link to libjevents. Check jevents is compiled with -fPIC.)"])
```
Comment (rubenk) [inline configure.ac]: I don't understand this. If libjevents is not compiled with -fPIC, it is not a shared library.

File: configure.ac [hunk 25]
```diff
@@ -4364,6 +4364,77 @@ then
...
+  CPPFLAGS="$SAVE_CPPFLAGS"
+  LDFLAGS="$SAVE_LDFLAGS"
+  LIBS="$SAVE_LIBS"
+fi
+if test "x$with_libjevents" = "xyes"
+then
+  BUILD_WITH_LIBJEVENTS_CPPFLAGS="$with_libjevents_cppflags"
+  BUILD_WITH_LIBJEVENTS_LDFLAGS="$with_libjevents_ldflags"
+  BUILD_WITH_LIBJEVENTS_LIBS="-ljevents"
+  AC_SUBST(BUILD_WITH_LIBJEVENTS_CPPFLAGS)
```
Comment (rubenk) [inline configure.ac]: You should not use AC_SUBST in a conditional statement. Please move it below the fi.

File: contrib/systemd.collectd.service (hunk 20 above)
Comment (serhiyx) [inline contrib/systemd.collectd.service]: This environment variable is not used by the plugin itself. It is used inside **jevents** library that is used by the plugin.

Comment (rubenk) [inline contrib/systemd.collectd.service]: And there's no way to pass this to jevents except from the environment?

File: configure.ac (hunk 23 above)
Comment (serhiyx) [inline configure.ac]: Actually I want to create shared object that is linked with jevents library just to check if jevents was compiled with -fPIC. I've added this logic to avoid errors during collectd installation/compilation - if collectd is compiled on the system with jevents installed we check if jevents version is suitable for us and if not just disable intel_pmu plugin during configure.

File: configure.ac (hunk 24 above)
Comment (serhiyx) [inline configure.ac]: By default jevents library is compiled without -fPIC. But this flag is required when jevents library is linked with pmu plugin (intel_pmu.so) to avoid the following error:

`/usr/bin/ld: /usr/local/lib64/libjevents.a(cache.o): relocation R_X86_64_32S against `.bss' can not be used when making a shared object; recompile with -fPIC`

File: configure.ac (hunk 23 above)
Comment (rubenk) [inline configure.ac]: I think it's more trouble than it's worth, you've already a note in the README that jevents should be position independent.

File: contrib/systemd.collectd.service (hunk 20 above)
Comment (serhiyx) [inline contrib/systemd.collectd.service]: Yes, it is possible. Will rework the solution to use config option and get rid of using of environment variable..

File: configure.ac (hunk 23 above)
Comment (serhiyx) [inline configure.ac]: Agree. Will remove this.

File: README [hunk 26]
```diff
@@ -240,7 +245,7 @@ Features
 
     - netapp
       Plugin to query performance values from a NetApp storage system using the
-      “Manage ONTAP” SDK provided by NetApp.
+      “Manage ONTAP” SDK provided by NetApp.
```
Comment (rubenk) [inline README]: What changed on this line?

File: configure.ac [hunk 27]
```diff
@@ -4364,6 +4364,53 @@ then
 fi
 # }}}
 
+# --with-libjevents {{{
+with_libjevents_cppflags=""
+with_libjevents_ldflags=""
+AC_ARG_WITH([libjevents],
+  [AS_HELP_STRING([--with-libjevents@<:@=PREFIX@:>@], [Path to libjevents.])],
+  [
+    if test "x$withval" != "xno" && test "x$withval" != "xyes"
```
Comment (rubenk) [inline configure.ac]: put if and then on the same line please.

File: configure.ac [hunk 28]
```diff
@@ -4364,6 +4364,53 @@ then
...
+  ],
+  [with_libjevents="yes"]
+)
+
+if test "x$with_libjevents" = "xyes"
+then
+  SAVE_CPPFLAGS="$CPPFLAGS"
+  CPPFLAGS="$CPPFLAGS $with_libjevents_cppflags"
+
+  AC_CHECK_HEADERS(jevents.h, [with_libjevents="yes"], [with_libjevents="no (jevents.h not found)"])
```
Comment (rubenk) [inline configure.ac]: please quote all the arguments.

File: configure.ac [hunk 29]
```diff
@@ -4364,6 +4364,53 @@ then
...
+
+  CPPFLAGS="$SAVE_CPPFLAGS"
+fi
+if test "x$with_libjevents" = "xyes"
+then
+  SAVE_LDFLAGS="$LDFLAGS"
+  LDFLAGS="$LDFLAGS $with_libjevents_ldflags"
+
+  AC_CHECK_LIB(jevents, json_events, [with_libjevents="yes"], [with_libjevents="no (Can't find libjevents)"])
```
Comment (rubenk) [inline configure.ac]: please quote all the arguments, and break this long line.

File: configure.ac [hunk 30]
```diff
@@ -4364,6 +4364,53 @@ then
...
+
+  LDFLAGS="$SAVE_LDFLAGS"
+fi
+if test "x$with_libjevents" = "xyes"
+then
+  BUILD_WITH_LIBJEVENTS_CPPFLAGS="$with_libjevents_cppflags"
+  BUILD_WITH_LIBJEVENTS_LDFLAGS="$with_libjevents_ldflags"
+  BUILD_WITH_LIBJEVENTS_LIBS="-ljevents"
+fi
+AC_SUBST(BUILD_WITH_LIBJEVENTS_CPPFLAGS)
```
Comment (rubenk) [inline configure.ac]: please quote the argument.

File: README (hunk 26 above)
Comment (serhiyx) [inline README]: must have been fixed by my editor. no-break space symbol (c2a0) was changed to space symbol (20).
Should I revert this change?

Comment (rubenk) [inline README]: Heh, nice one. No problem, please leave it in, I prefer spaces anyway :)

File: Makefile.am [hunk 31]
```diff
@@ -876,6 +876,14 @@ hugepages_la_SOURCES = src/hugepages.c
 hugepages_la_LDFLAGS = $(PLUGIN_LDFLAGS)
 endif
 
+if BUILD_PLUGIN_INTEL_PMU
+pkglib_LTLIBRARIES += intel_pmu.la
+intel_pmu_la_SOURCES = src/intel_pmu.c
+intel_pmu_la_CFLAGS = $(AM_CFLAGS) $(BUILD_WITH_LIBJEVENTS_CPPFLAGS)
```
Comment (rubenk) [inline Makefile.am]: BUILD_WITH_LIBJEVENTS_CPPFLAGS should be added to intel_pmu_la_CPPFLAGS.

File: src/intel_pmu.c [hunk 32]
```diff
@@ -0,0 +1,577 @@
...
+  for (e = g_ctx.event_list->eventlist; e; e = e->next) {
+    DEBUG(PMU_PLUGIN ":   event       : %s", e->event);
+    DEBUG(PMU_PLUGIN ":     group_lead: %d", e->group_leader);
+    DEBUG(PMU_PLUGIN ":     end_group : %d", e->end_group);
+    DEBUG(PMU_PLUGIN ":     type      : %#x", e->attr.type);
+    DEBUG(PMU_PLUGIN ":     config    : %#x", (unsigned)e->attr.config);
+    DEBUG(PMU_PLUGIN ":     size      : %d", e->attr.size);
+  }
+
+  return;
```
Comment (rubenk) [inline src/intel_pmu.c]: no need to return from a void function.

File: src/intel_pmu.c [hunk 33]
```diff
@@ -0,0 +1,577 @@
...
+  DEBUG(PMU_PLUGIN ": Config:");
+  DEBUG(PMU_PLUGIN ":   hw_cache_events   : %d", g_ctx.hw_cache_events);
+  DEBUG(PMU_PLUGIN ":   kernel_pmu_events : %d", g_ctx.kernel_pmu_events);
+  DEBUG(PMU_PLUGIN ":   software_events   : %d", g_ctx.sw_events);
+
+  for (size_t i = 0; i < g_ctx.hw_events_count; i++) {
+    DEBUG(PMU_PLUGIN ":   hardware_events[%zu]: %s", i, g_ctx.hw_events[i]);
+  }
+
+  return;
```
Comment (rubenk) [inline src/intel_pmu.c]: no need to return from a void function.

File: src/intel_pmu.c [hunk 34]
```diff
@@ -0,0 +1,577 @@
...
+    }
+
+    g_ctx.hw_events_count++;
+  }
+
+  return 0;
+}
+
+static int pmu_config(oconfig_item_t *ci) {
+  int ret = 0;
```
Comment (rubenk) [inline src/intel_pmu.c]: can you reduce the scope by moving this into the for loop?

File: src/intel_pmu.c [hunk 35]
```diff
@@ -0,0 +1,577 @@
...
+    } else if (strcasecmp("EventList", child->key) == 0) {
+      ret = cf_util_get_string_buffer(child, g_ctx.event_list_fn,
+                                      sizeof(g_ctx.event_list_fn));
+    } else if (strcasecmp("HardwareEvents", child->key) == 0) {
+      ret = pmu_config_hw_events(child);
+    } else if (strcasecmp("ReportSoftwareEvents", child->key) == 0) {
+      ret = cf_util_get_boolean(child, &g_ctx.sw_events);
+    } else {
+      ERROR(PMU_PLUGIN ": Unknown configuration parameter \"%s\".", child->key);
+      ret = (-1);
```
Comment (rubenk) [inline src/intel_pmu.c]: no need for the braces.

File: src/intel_pmu.c [hunk 36]
```diff
@@ -0,0 +1,577 @@
...
+  } else {
+    ssnprintf(vl.plugin_instance, sizeof(vl.plugin_instance), "%d", cpu);
+  }
+  sstrncpy(vl.type, "counter", sizeof(vl.type));
+  sstrncpy(vl.type_instance, event, sizeof(vl.type_instance));
+
+  plugin_dispatch_values(&vl);
+}
+
+static int pmu_dispatch_data(void) {
```
Comment (rubenk) [inline src/intel_pmu.c]: this function always returns 0, make it void?

File: src/intel_pmu.c [hunk 37]
```diff
@@ -0,0 +1,577 @@
...
+  DEBUG(PMU_PLUGIN ": %s:%d", __FUNCTION__, __LINE__);
+
+  ret = read_all_events(g_ctx.event_list);
+  if (ret != 0) {
+    ERROR(PMU_PLUGIN ": Failed to read values of all events.");
+    return 0;
+  }
+
+  ret = pmu_dispatch_data();
+  if (ret != 0) {
```
Comment (rubenk) [inline src/intel_pmu.c]: this is a no-op, pmu_dispatch_data always returns 0.

File: src/intel_pmu.c [hunk 38]
```diff
@@ -0,0 +1,577 @@
...
+
+static int pmu_read(__attribute__((unused)) user_data_t *ud) {
+  int ret;
+
+  DEBUG(PMU_PLUGIN ": %s:%d", __FUNCTION__, __LINE__);
+
+  ret = read_all_events(g_ctx.event_list);
+  if (ret != 0) {
+    ERROR(PMU_PLUGIN ": Failed to read values of all events.");
+    return 0;
```
Comment (rubenk) [inline src/intel_pmu.c]: why not return an error?

File: src/intel_pmu.c [hunk 39]
```diff
@@ -0,0 +1,577 @@
...
+  }
+
+  ret = pmu_dispatch_data();
+  if (ret != 0) {
+    ERROR(PMU_PLUGIN ": Failed to dispatch event values.");
+    return 0;
```
Comment (rubenk) [inline src/intel_pmu.c]: why not return an error?

File: src/intel_pmu.c [hunk 40]
```diff
@@ -0,0 +1,577 @@
...
+        calloc(sizeof(struct event) + sizeof(struct efd) * el->num_cpus, 1);
+    if (e == NULL) {
+      ERROR(PMU_PLUGIN ": Failed to allocate event structure");
+      return -ENOMEM;
+    }
+
+    e->attr.type = type;
+    e->attr.config = events[i].config;
+    e->attr.size = PERF_ATTR_SIZE_VER0;
+    e->next = NULL;
```
Comment (rubenk) [inline src/intel_pmu.c]: this is a no-op, the struct has already been zeroed by calloc.

File: src/intel_pmu.c [hunk 41]
```diff
@@ -0,0 +1,577 @@
...
+  if (ret != 0) {
+    ERROR(PMU_PLUGIN ": Failed to dispatch event values.");
+    return 0;
+  }
+
+  return 0;
+}
+
+static int pmu_add_events(struct eventlist *el, uint32_t type,
+                          event_info_t *events, int count) {
```
Comment (rubenk) [inline src/intel_pmu.c]: count should be size_t.

**Comment by rubenk** [comment]
@serhiyx I left a few more review comments. Once these are resolved I think this is good to go in. Feel free to also add yourself to AUTHORS if you want to.

**Comment by rubenk** [comment]
@octo I think this is good to go in, and think all your remarks have been addressed. Do you want to merge this?

**Comment by octo** [comment]
Thank you very much for your great work, @serhiyx!
